packages/core/src/scp-cf/destination-accessor.ts

```diff
@@ -482,19 +285,346 @@ const emptyDestinationByType: DestinationsByType = {
   subaccount: []
 };
 
-function subscriberDestinationIsSelected(
-  selectionStrategy: DestinationSelectionStrategy
-): boolean {
-  return (
-    selectionStrategy === subscriberFirst ||
-    selectionStrategy === alwaysSubscriber
-  );
-}
-
 function emptyDestinationsByType(
   destinationByType: DestinationsByType
 ): boolean {
   return (
-    !destinationByType.instance.length && !destinationByType.instance.length
+    !destinationByType.instance.length && !destinationByType.subaccount.length
   );
 }
+
+class DestinationAccessor {
+  public static async getDestinationFromDestinationService(
+    name: string,
+    options: DestinationOptions & { iss?: string }
+  ): Promise<Destination | undefined> {
+    const decodedUserJwt = await DestinationAccessor.getDecodedUserJwt(options);
+    const providerToken = await DestinationAccessor.getProviderToken(options);
+    const da = new DestinationAccessor(
+      name,
+      options,
+      decodedUserJwt,
+      providerToken
+    );
+
+    return (
+      da.trySubscriberCache() ??
+      (await da.tryProviderCache()) ??
+      (await da.processDestinationResult(await da.tryForOAuth())) ??
+      (await da.processDestinationResult(await da.tryForCert())) ??
+      (await da.processDestinationResult(await da.tryService())) ??
+      undefined
+    );
+  }
+
+  private static async getDecodedUserJwt(
+    options: DestinationOptions & { iss?: string }
+  ): Promise<DecodedJWT | undefined> {
+    return options.userJwt
+      ? verifyJwt(options.userJwt, options)
+      : options.iss
+      ? { iss: options.iss }
+      : undefined;
+  }
+
+  private static async getProviderToken(
+    options: DestinationOptions
+  ): Promise<string> {
+    const { userJwt, ...optionsWithoutJwt } = options;
+    return serviceToken('destination', optionsWithoutJwt);
+  }
+
+  private subscriberDestinationCache: DestinationsByType | undefined;
+  private providerDestinationCache: DestinationsByType | undefined;
```

Comment:
I have to say that these two variables are confusing me immensely.
Both `trySubscriberCache` as well as `tryProviderCache` are reading from the actual `destinationCache` that is defined outside of this class, and since the tests are passing I'm guessing that cross-request caching still works (unless the tests are broken, of course). 
If that is the case, then what are these variables doing? They're called `somethingCache`, but they're not actually caches (as also indicated by the type). From how they're used it looks like you're trying to prevent that the same HTTP requests are executed multiple times in the scope of a single conceptual "hey gimme destination X" operation. But how would that happen in the first place?

Reply:
I also was not happy with the name. These variables are used so that the services are called a minimal amount of times for one fetch round trip. If the cache destination is switched on they are useless. But  if the cache is switched of by the config it prevents the code to be highly inefficient compared to the old implementation, where all was in one method so that also without a cache only one call was done.

Reply:
Okay I get the reasoning, but I still don't understand this: In which scenario in the old code would the same endpoint have been called twice?